server/auth.go

```diff
@@ -78,15 +89,65 @@ func (h *Handler) LoginPost(c echo.Context) error {
 }
 
 func (h *Handler) LoginGet(c echo.Context) error {
+	sess, _ := session.Get(CookieKeySession, c)
+
 	vars := map[string]interface{}{
 		"csrf":               c.Get("csrf").(string),
 		"globus":             viper.GetBool("globus_signup"),
 		"enable_user_signup": viper.GetBool("enable_user_signup"),
+		"success":            sess.Flashes("success"),
 	}
 
+	sess.Save(c.Request(), c.Response())
+
 	return c.Render(http.StatusOK, "login.html", vars)
 }
 
+func (h *Handler) ChangePost(c echo.Context) error {
+	message := ""
+	sess, _ := session.Get(CookieKeySession, c)
+
+	uid := c.FormValue("uid")
+	current := c.FormValue("password")
+	pass := c.FormValue("new_password")
+	pass2 := c.FormValue("new_password2")
+	challenge := c.FormValue("challenge")
+
+	if pass != pass2 {
+		message = "Passwords do not match!"
+	} else {
+		err := h.setPassword(uid, current, pass, challenge)
+		if err != nil {
+			message = err.Error()
+		} else {
+			sess.AddFlash("Password successfully changed!", "success")
+			sess.Save(c.Request(), c.Response())
+			return c.Redirect(http.StatusFound, Path("/auth/login"))
+		}
+	}
+
+	vars := map[string]interface{}{
+		"csrf":               c.Get("csrf").(string),
+		"globus":             viper.GetBool("globus_signup"),
+		"enable_user_signup": viper.GetBool("enable_user_signup"),
+		"message":            message,
+		"uid":                uid,
+	}
+
+	return c.Render(http.StatusOK, "change.html", vars)
+}
+
+func (h *Handler) ChangeGet(c echo.Context) error {
+	vars := map[string]interface{}{
+		"csrf":               c.Get("csrf").(string),
+		"globus":             viper.GetBool("globus_signup"),
+		"enable_user_signup": viper.GetBool("enable_user_signup"),
+		"uid":                c.QueryParam("uid"),
+	}
+
```

Comment:
We need to check cookie here to ensure user authenticated with an expired password, otherwise return 401.